iroh-gossip/examples/chat.rs

```diff
@@ -0,0 +1,345 @@
+    bind_port: u16,
+    #[clap(subcommand)]
+    command: Command,
+}
+
+#[derive(Parser, Debug)]
+enum Command {
+    OpenTopic { topic: String },
+    JoinTicket { ticket: String },
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    tracing_subscriber::fmt::init();
+    let args = Args::parse();
+
+    // parse or generate our keypair
+    let keypair = match args.private_key {
+        None => Keypair::generate(),
+        Some(key) => parse_keypair(&key)?,
+    };
+    println!("> our private key: {}", fmt_secret(&keypair));
+
+    // configure our derp map
+    let derp_map = match (args.no_derp, args.derp) {
+        (false, None) => Some(default_derp_map()),
+        (false, Some(url)) => Some(derp_map_from_url(url)?),
+        (true, None) => None,
+        (true, Some(_)) => bail!("You cannot set --no-derp and --derp at the same time"),
+    };
+    println!("> using DERP servers: {:?}", fmt_derp_map(&derp_map));
+
+    // init a cell that will hold our gossip handle to be used in endpoint callbacks
+    let gossip_cell: OnceCell<GossipHandle> = OnceCell::new();
+    // init a channel that will emit once the initial endpoints of our local node are discovered
+    let (initial_endpoints_tx, mut initial_endpoints_rx) = mpsc::channel(1);
+
+    // build our magic endpoint
+    let gossip_cell_clone = gossip_cell.clone();
+    let endpoint = MagicEndpoint::builder()
+        .keypair(keypair)
+        .alpns(vec![GOSSIP_ALPN.to_vec()])
+        .derp_map(derp_map)
+        .on_endpoints(Box::new(move |endpoints| {
+            // send our updated endpoints to the gossip protocol to be sent as PeerData to peers
+            if let Some(gossip) = gossip_cell_clone.get() {
+                gossip.update_endpoints(endpoints).ok();
+            }
+            // trigger oneshot on the first endpoint update
+            initial_endpoints_tx.try_send(endpoints.to_vec()).ok();
+        }))
+        .bind(args.bind_port)
+        .await?;
+    println!("> our peer id: {}", endpoint.peer_id());
+
+    // wait for a first endpoint update so that we know about at least one of our addrs
+    let initial_endpoints = initial_endpoints_rx.recv().await.unwrap();
+    println!("> our endpoints: {initial_endpoints:?}");
+
+    // create the gossip protocol
+    let gossip = GossipHandle::from_endpoint(endpoint.clone(), Default::default());
+    // insert the gossip handle into the gossip cell to be used in the endpoint callbacks above
+    gossip_cell.set(gossip.clone()).unwrap();
+
+    // pass our initial peer info to the gossip protocol
+    gossip.update_endpoints(&initial_endpoints)?;
+
+    // spawn our endpoint loop that forwards incoming connections to the gossiper
+    tokio::spawn(endpoint_loop(endpoint.clone(), gossip.clone()));
+
+    let mut our_ticket = match &args.command {
+        Command::OpenTopic { topic } => {
+            println!("> opening topic {topic} and waiting for peers to join us...");
```

Comment:
this within the example is confusing. The topic passed to the example is a `utf-8` string, but the real topic is a `hash`. It would be far clearer to print the topic hash along the topic name, since that's what's actually printed on the join side.

Right now I get for example:
```bash
$ <command> <args> open-topic testing
# > trying opening topic testing and waiting for peers to join us...
# > ticket to join us: mhgjrzbn5wlia6agx4lcbyj4jzvbxbigrswzgobkfyyqpqtjv37aa
```

and then on the joining side:
```bash
$ <command> <args> join-ticket mhgjrzbn5wlia6agx4lcbyj4jzvbxbigrswzgobkfyyqpqtjv37aa
# > joining topic mhgjrzbn5wlia6agx4lcbyj4jzvbxbigrswzgobkfyyqpqtjv37a and connecting to []
```

someone trying this example would in principle expect `testing` on the joining side. If this is (likely) not possible to get back, giving the hash on the `topic` command makes this match

Reply:
Oh yes, this is a good point. Thanks for raising, not too much thought has gone into the semantics of this example.
With the most recent comment, I removed the topic-is-a-hashed-string "feature". Instead, you now either pass in a topic id directly as 32 bytes in base32 encoding, or a random topic id will be created.
I think it is much clearer this way for the example.